src/rules.jl

```diff
@@ -96,3 +96,112 @@ true
 See also: [`frule`](@ref), [`@scalar_rule`](@ref)
 """
 rrule(::Any, ::Vararg{Any}; kwargs...) = nothing
+
+
+#######################################################################
+# Infastructure to support generating overloads from rules.
+
+const NEW_RRULE_HOOKS = Function[]
+const NEW_FRULE_HOOKS = Function[]
+_hook_list(::typeof(rrule)) = NEW_RRULE_HOOKS
+_hook_list(::typeof(frule)) = NEW_FRULE_HOOKS
+
+"""
+    on_new_rule(sig->eval...), frule | rrule)
+"""
+function on_new_rule(hook_fun, rule_kind)
+    # get all the existing rules
+    ret = map(_rule_list(rule_kind)) do method
+        sig = _primal_sig(rule_kind, method)
+        _safe_hook_fun(hook_fun, sig)
+    end
+
+    # register hook for new rules
+    push!(_hook_list(rule_kind), hook_fun)
+    return ret
+end
+
+function __init__()
+    # Need to refresh rules when a module is loaded or a file is `include`d.
+    push!(Base.package_callbacks, pkgid -> refresh_rules())
+    push!(Base.include_callbacks, (mod, filename) -> refresh_rules())
+end
```

Comment:
Possible we should also hook to [`Revise.entr`](https://timholy.github.io/Revise.jl/stable/user_reference/#Revise.entr) here.
I think that would be:
```julia
Revise.entr(refresh_rules, []; all=true, postpone=true)
```
Having that would make sure that modifications to rules are picked up to be regenerated,
but it would not handle deletions.
We would need to wrap it in Requires.jl.

I think this is best left for a follow up PR, and an issue should be openned about it.
Might also warrent a chat with Tim Holy, see if he has bright ideas about how we can handle deletions